Unit test failure on high CPU core count machines

<details><summary>Click to expand!</summary> 
 
 ### Issue Type

Bug

### Source

source

### Tensorflow Version

git HEAD

### Custom Code

No

### OS Platform and Distribution

CentOS 7

### Mobile device

n/a

### Python version

3.8.10

### Bazel version

5.1.1

### GCC/Compiler version

10.2.1

### CUDA/cuDNN version

n/a

### GPU model and memory

n/a

### Current Behaviour?

```shell
//tensorflow/python/data/experimental/kernel_tests/service:cross_trainer_cache_test fails if there are more than 48 CPU cores in the machine being used to test.
```


### Standalone code to reproduce the issue

```shell
bazel test --test_timeout=300,500,-1,-1 --flaky_test_attempts=2 --test_output=all --cache_test_results=no --config=nonccl --copt=-mtune=generic --copt=-march=armv8-a --copt=-O3 --verbose_failures --build_tag_filters=-no_oss,-oss_serial,-gpu,-tpu,-benchmark-test,-v1only,-no_aarch64,-requires-gpu --test_tag_filters=-no_oss,-oss_serial,-gpu,-tpu,-benchmark-test,-v1only,-no_aarch64,-requires-gpu --build_tests_only -- //tensorflow/python/data/experimental/kernel_tests/service:cross_trainer_cache_test
```


### Relevant log output

```shell
FAIL: testConcurrentReaders_test_mode_graph_tfapiversion_2 (__main__.CrossTrainerCacheTest)
CrossTrainerCacheTest.testConcurrentReaders_test_mode_graph_tfapiversion_2
testConcurrentReaders_test_mode_graph_tfapiversion_2(mode='graph', tf_api_version=2)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/builder/.cache/bazel/_bazel_builder/9dc2dbd69dc3512cedb530e1521082e7/execroot/org_tensorflow/bazel-out/aarch64-opt/bin/tensorflow/python/data/experimental/kernel_tests/service/cross_trainer_cache_test.runfiles/absl_py/absl/testing/parameterized.py", line 314, in bound_param_test
    return test_method(self, **testcase_params)
  File "/home/builder/.cache/bazel/_bazel_builder/9dc2dbd69dc3512cedb530e1521082e7/execroot/org_tensorflow/bazel-out/aarch64-opt/bin/tensorflow/python/data/experimental/kernel_tests/service/cross_trainer_cache_test.runfiles/org_tensorflow/tensorflow/python/framework/test_combinations.py", line 362, in decorated
    execute_test_method()
  File "/home/builder/.cache/bazel/_bazel_builder/9dc2dbd69dc3512cedb530e1521082e7/execroot/org_tensorflow/bazel-out/aarch64-opt/bin/tensorflow/python/data/experimental/kernel_tests/service/cross_trainer_cache_test.runfiles/org_tensorflow/tensorflow/python/framework/test_combinations.py", line 345, in execute_test_method
    test_method(**kwargs_to_pass)
  File "/home/builder/.cache/bazel/_bazel_builder/9dc2dbd69dc3512cedb530e1521082e7/execroot/org_tensorflow/bazel-out/aarch64-opt/bin/tensorflow/python/data/experimental/kernel_tests/service/cross_trainer_cache_test.runfiles/org_tensorflow/tensorflow/python/data/experimental/kernel_tests/service/cross_trainer_cache_test.py", line 97, in testConcurrentReaders
    self.assertEqual(self.evaluate(iterators[j]()), i)
AssertionError: 9 != 0

----------------------------------------------------------------------
```
</details>

--- a/tensorflow/python/data/experimental/kernel_tests/service/cross_trainer_cache_test.py
+++ b/tensorflow/python/data/experimental/kernel_tests/service/cross_trainer_cache_test.py
@@ -24,6 +24,8 @@ from tensorflow.python.framework import combinations
 from tensorflow.python.framework import errors
 from tensorflow.python.platform import test
 
+import multiprocessing
+
 
 class CrossTrainerCacheTest(data_service_test_base.TestBase,
                             parameterized.TestCase):
@@ -71,8 +73,18 @@ class CrossTrainerCacheTest(data_service_test_base.TestBase,
       combinations.times(
           combinations.combine(tf_api_version=2, mode=["eager", "graph"])))
   def testConcurrentReaders(self):
+    # Fetching an element from the dataset will trigger prefetches of more
+    # elements, one per CPU core which will be placed in the cache.
+    # However if the number of prefetches exceeds the space available in
+    # the cache then the sliding window will be moved forward away from
+    # the element just read thus negating the use of the cache as other
+    # trainers will not get the correct element.
+    # Hence the need to calculate the size of the cache based on the
+    # number of CPU cores and the element size of 363. The extra 8
+    # entries are simply a bit of margin.
+    num_cpus = multiprocessing.cpu_count()
     cluster = self._create_cluster(
-        num_workers=1, cross_trainer_cache_size_bytes=18000)
+        num_workers=1, cross_trainer_cache_size_bytes=(num_cpus + 8) * 363)
     num_readers = 20
     num_elements = 50
     dataset = dataset_ops.Dataset.range(10000000).repeat()

